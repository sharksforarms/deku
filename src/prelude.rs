pub use crate::{error::DekuError, BitsReader, BitsWriter, DekuRead, DekuWrite};
pub use bitvec::{
    order::BitOrder, order::Lsb0, order::Msb0, slice::AsBits, slice::BitSlice, vec::BitVec,
};
