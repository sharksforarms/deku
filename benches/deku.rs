use criterion::{black_box, criterion_group, criterion_main, Criterion};
use deku::prelude::*;

#[derive(Debug, PartialEq, DekuRead)]
struct DekuByte {
    data: u8
}

#[derive(Debug, PartialEq, DekuRead)]
#[deku(id_type = "u8")]
enum DekuEnum {
    #[deku(id = "0x01")]
    VariantA(u8),
}

#[derive(Debug, PartialEq, DekuRead)]
struct DekuVec {
    count: u8,
    #[deku(len = "count")]
    data: Vec<u8>
}

fn deku_read_byte(input: &[u8]) {
    let (_rest, v) = DekuByte::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuByte { data: 0x01 }, v);
}

fn deku_read_enum(input: &[u8]) {
    let (_rest, v) = DekuEnum::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuEnum::VariantA(0x02), v);
}

fn deku_read_vec(input: &[u8]) {
    let (_rest, v) = DekuVec::from_bytes((input, 0)).unwrap();
    assert_eq!(100, v.count);
    assert_eq!(100, v.data.len());
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("deku_read_byte", |b| {
        b.iter(|| deku_read_byte(black_box([0x01].as_ref())))
    });

    c.bench_function("deku_read_enum", |b| {
        b.iter(|| deku_read_enum(black_box([0x01, 0x02].as_ref())))
    });

    let deku_read_vec_input = {
        let mut v = [0xFFu8; 101].to_vec();
        v[0] = 100u8;
        v
    };
    c.bench_function("deku_read_vec", |b| {
        b.iter(|| deku_read_vec(black_box(&deku_read_vec_input)))
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
