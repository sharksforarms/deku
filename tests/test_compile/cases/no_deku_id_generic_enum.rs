use deku::prelude::*;

#[derive(DekuRead)]
#[deku(id_type = "u8")]
pub enum Body<T: for<'a> DekuReader<'a>> {
    #[deku(id = "0x0001")]
    First(T),
}

fn main() {
    let n = Body::<u8>::First(1);
    n.deku_id();
}
