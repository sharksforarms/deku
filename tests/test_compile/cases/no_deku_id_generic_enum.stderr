error[E0599]: no method named `deku_id` found for enum `Body` in the current scope
  --> tests/test_compile/cases/no_deku_id_generic_enum.rs:12:7
   |
5  | pub enum Body<T: for<'a> DekuReader<'a>> {
   | ---------------------------------------- method `deku_id` not found for this enum
...
12 |     n.deku_id();
   |       ^^^^^^^ method not found in `Body<u8>`
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `deku_id`, perhaps you need to implement it:
           candidate #1: `deku::DekuEnumExt`
